### ISSUE TITLE
Error when deserialising an enum within a tagged enum.

### ISSUE THREAD

**Issue by 46bit** [serde-rs/serde#775]
I've found a bug where a tagged enum can be serialised but not deserialised. Here's a testcase ([runnable gist here](https://gist.github.com/46bit/4f9a2daaa8a32cd27a5e7a39bf00154a)):

``` rust
#[derive(PartialEq, Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "tag")]
pub enum A {
    A1 { field: B },
    A2,
}

#[derive(PartialEq, Clone, Debug, Serialize, Deserialize)]
pub enum B {
    B1,
    B2,
}

fn main() {
    // Construct an A containing a B.
    let a = A::A1 { field: B::B1 };
    println!("{:?}", a);

    // Successfully serialises to `{"tag":"A1","field":"B1"}`.
    let json = serde_json::to_string(&a).unwrap();
    println!("{}", json);

    // That correct, serde_json-generated JSON errors in the decode with
    // `Syntax(Message("invalid type: string \"B1\", expected enum B"), 0, 0)`
    let a_new: A = serde_json::from_str(json.as_str()).unwrap();
    println!("{:?}", a_new);
}
```

If you remove `#[serde(tag = "tag")]` (or, I think, change it to be untagged) then that format deserialises fine.

**Comment by 46bit** [comment]
Hi. Thanks to guidance from @dtolnay I've been working on a `Content(Ref)Deserializer::deserialize_enum` implementation to fix this bug. 

I'm not sure how to convert something. `serde_json::Value::Map<String, Value>` restricts keys to be `String`. But `serde::de::Content::Map<Content, Content>` does not. Should I expect keys to be `String` or return an `Unexpected`, or is there something else I should do?

Current progress: https://github.com/46bit/serde/commit/8f6e1fb5

**Comment by dtolnay** [comment]
The EnumDeserializer needs to store Content instead of String and then use `seed.deserialize(self.variant)` instead of using String::into_deserializer.

**Comment by dtolnay** [comment]
Reopening until we fix this for the untagged case as well - might as well track them together.

**Comment by dtolnay** [comment]
Fixed in serde 0.9.9.
